lrpar/cttests/src/multiparam.test

```diff
@@ -0,0 +1,27 @@
+name: Test multiple types
+yacckind: Grmtools
+grammar: |
+    %parse_param <'a>(_a_count: u64, _b_count: u64)
+    %start S
+    %%
+    S  -> Result<(), ()>
+    : AorB   { Ok($1?) };
+
+    AorB -> Result<(), ()>
+    :        { Ok(())  }
+    | AorB B { Ok($2?) }
+    | AorB A { Ok($2?) };
+
+    A  -> Result<(), ()>
+    : 'a'    { Ok(*_a_count.deref_mut() += 1) };
+
+    B  -> Result<(), ()>
+    : 'b'    { Ok(*_b_count.deref_mut() += 1) };
+    %%
+    // We get &mut &mut u64's, use deref_mut to coerce them to &mut u64
```

Comment:
This slightly surprised me: naively, I specified `u64` to %parse_param, so I would have expected `u64`, not `&mut u64` and almost certainly not `&mut &mut u64`. Clearly I've misunderstood something somewhere, but I'm not quite sure where :)

Reply:
*nod*, this part that bothers me a lot too, I really should have brought more attention to it earlier. The problem is we have to pass the tuple by reference, while avoiding moving the tuple.

Here is a mildly reflective playground link, the only difference is `(x, y)` vs  `&mut (x, y)`, of how we destructure the pattern. [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cec88192e245564e61d7b9b378ef43e1)

Normally we'd expect something like the `move` variants, where we'd destructure the tuple via `&mut (x, y)` giving us an `x`, and `y` of `&u64`, what works here though is the `nomove`, which ends up adding the tuples own reference onto the existing reference so we end up with the double indirection.

With the mechanism where we return the ParamT from the actions, we can move the tuple, because wherever we move it to should give us back the place that they moved it.

It seems like a place where we are cornered between two fairly ugly choices. :face_with_head_bandage:

Reply:
The second part that bothers me about this, is we have to make a choice for the user between `&mut`, and `&` ref, because there is no mutation polymorphic reference, with unique ownership.

Reply:
In conclusion I *think* perhaps the best thing to do here, is just only accept a single parameter, meaning the parse caller has to tuplize multiple params themself.

Reply:
OK, I'm starting to see the problem. This works:

```rust
fn f(t: &mut (&mut u64, &mut u64)) {
    *t.0 = 4;
}

fn main() {
    let mut x = 2;
    let mut y = 3;
    f(&mut (&mut x, &mut y));
}
```

but this doesn't:

```rust
fn f(t: &(&mut u64, &mut u64)) {
    *t.0 = 4;
}

fn main() {
    let mut x = 2;
    let mut y = 3;
    f(&mut (&mut x, &mut y));
}
```

However, this does work:

```rust
fn f(t: (&mut u64, &mut u64)) {
    *t.0 = 4;
}

fn main() {
    let mut x = 2;
    let mut y = 3;
    f((&mut x, &mut y));
    println!("{}", x);
}
```

I'm not sure I have any deep thoughts on this yet, but at least I'm starting to see what the design issue is, and I agree that it's not an ideal one. However, I'm an optimist -- I'm sure we'll find a good solution one way or another!

Reply:
Indeed it does, i'm pretty bad at traits, but it *seems* like we might be able to do something with those having the action functions implement trait which works for some fixed number of tuples, and as well as refs.

Reply:
I could definitely live with the old "works for tuples up to size N" hack -- it isn't pretty, but it is effective (and that trick is used elsewhere in Rust, so it does have precedent)!

Reply:
Yeah seems like it might work, not sure how much of e.g. `Parser` would need to be parameterized across the various tuples though.

unfortunately, it looks as though due to some hardware issues, it might be some time before I can do any further work here, I'll try and get back to it once I have a new development machine up and running.

Reply:
> Yeah seems like it might work, not sure how much of e.g. Parser would need to be parameterized across the various tuples though.

That's a good point. The more I think about it, the more I think that if we take the "traditional" approach to parsers and generate all the code we'd avoid all of these problems, but that isn't a 5 minute job.

One possibility is that we could merge this work more-or-less as-is and either-or a) not document the feature b) explicitly say that it's going to change (in a more ergonomic way) in the future. Any thoughts?

If it would help to have access to a shell machine for development, please drop me an email, and I can sort something out!